exonum/src/messages2/mod.rs

```diff
@@ -0,0 +1,283 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Handling messages received from P2P node network.
+//!
+//! Every message passes through three phases:
+//!
+//!   * `Vec<u8>`: raw bytes as received from the network
+//!   * `SignedMessage`: integrity and signature of the message has been verified
+//!   * `Protocol`: the message has been completely parsed and has correct structure
+//!
+//! Graphical representation of the message processing flow:
+//!
+//! ```text
+//! +---------+             +---------------+                  +----------+
+//! | Vec<u8> |--(verify)-->| SignedMessage |--(deserialize)-->| Protocol |-->(handle)
+//! +---------+     |       +---------------+        |         +----------+
+//!                 |                                |
+//!                 V                                V
+//!              (drop)                           (drop)
+//! ```
+
+use byteorder::{ByteOrder, LittleEndian};
+use failure::Error;
+use hex::{FromHex, ToHex};
+
+use std::{borrow::Cow, cmp::PartialEq, fmt, mem, ops::Deref};
+
+use crypto::{hash, CryptoHash, Hash, PublicKey};
+use encoding;
+use storage::StorageValue;
+
+pub(crate) use self::{authorization::SignedMessage, helpers::HexStringRepresentation};
+pub use self::{helpers::BinaryForm, protocol::*};
+
+#[macro_use]
+mod compatibility;
+mod authorization;
+mod helpers;
+mod protocol;
+#[cfg(test)]
+mod tests;
+
+/// Version of the protocol. Different versions are incompatible.
+pub const PROTOCOL_MAJOR_VERSION: u8 = 1;
+
+/// Transaction raw buffer.
+/// This struct is used to transfer transactions in network.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct RawTransaction {
+    service_id: u16,
+    transaction_set: ServiceTransaction,
```

Comment:
Why can't we have 
```
struct RawTransaction:
  service_id : u16
  tx_id: u16
  payload: bytes[]
```
?

Reply:
So, given I have a `service_id` of a tx receiver, would an API to submit transaction look like this:

```
submitTransaction(receiver_id: u16, tx_parameters: ServiceTransaction) -> Hash
```